File: frontend/src/Components/Requests/AdminVideoRequestApprovalRequest.js
import React, { useEffect, useState } from 'react';
import server from '../Static/Constants';

const AdminVideoRequestApprovalRequest = () => {
    const [statuses, setStatuses] = useState([]);
    const [requestData, setRequestData] = useState({});
    const [loading, setLoading] = useState(true);

    const authHeaders = {
        "Authorization": localStorage.getItem("token"),
        "Admin-Token": localStorage.getItem("adminToken"),
    };

    const loadStatuses = async () => {
        try {
            const res = await fetch(`${server}/admins/video_request_options/`, {
                method: "GET",
                headers: authHeaders
            });

            const data = await res.json();
            const options = data?.data || [];
            setStatuses(options);
            return options;
        } catch (err) {
            console.error("Failed to load status options:", err);
            return [];
        }
    };

    const loadRequestsForStatus = async (status) => {
        try {
            const res = await fetch(`${server}/admins/video_requests/${status}/`, {
                method: "GET",
                headers: authHeaders
            });

            const data = await res.json();
            return Object.values(data?.data || {});
        } catch (err) {
            console.error(`Failed to load requests for ${status}:`, err);
            return [];
        }
    };

    useEffect(() => {
        const init = async () => {
            setLoading(true);

            const options = await loadStatuses();

            const allRequestData = {};
            for (const status of options) {
                allRequestData[status] = await loadRequestsForStatus(status);
            }

            setRequestData(allRequestData);
            setLoading(false);
        };

        init();
    }, []);

    const updateStatus = async (serial, newStatus) => {
        try {
            await fetch(`${server}/admins/review_video_request/${serial}/`, {
                method: "PATCH",
                headers: {
                    ...authHeaders,
                    "Content-Type": "application/json",
                },
                body: JSON.stringify({
                    request_status: newStatus
                }),
            });

            const refreshedData = {};
            for (const status of statuses) {
                refreshedData[status] = await loadRequestsForStatus(status);
            }

            setRequestData(refreshedData);

        } catch (err) {
            console.error("Failed to update status:", err);
        }
    };

    if (loading) return <p>Loading...</p>;

    return (
        <div style={{ display: "flex", gap: "20px", padding: "20px" }}>
            {statuses.map((status) => (
                <div key={status} style={{ width: "300px" }}>
                    <h2>{status}</h2>

                    <div style={{ border: "1px solid #aaa", padding: "10px" }}>
                        {requestData[status]?.length === 0 && (
                            <p>No requests in this status.</p>
                        )}

                        {requestData[status]?.map((req) => (
                            <div
                                key={req.serial}
                                style={{
                                    marginBottom: "15px",
                                    padding: "5px",
                                    borderBottom: "1px solid #ccc"
                                }}
                            >
                                <p><b>{req.requeset_title}</b></p>
                                <p>User: {req.username}</p>
                                <p>Date: {req.request_date}</p>
                                <p>{req.request_description}</p>

                                <select
                                    value={status}
                                    onChange={(e) =>
                                        updateStatus(req.serial, e.target.value)
                                    }
                                >
                                    {statuses.map((s) => (
                                        <option key={s} value={s}>
                                            {s}
                                        </option>
                                    ))}
                                </select>
                            </div>
                        ))}
                    </div>
                </div>
            ))}
        </div>
    );
};

export default AdminVideoRequestApprovalRequest;

File: flutter_frontend/lib/assets/admin/requests/adminVideoRequestApprovalRequest.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:flutter_frontend/static/constants.dart';

class AdminVideoRequestApprovalRequest extends StatefulWidget {
  const AdminVideoRequestApprovalRequest({super.key});

  @override
  State<AdminVideoRequestApprovalRequest> createState() =>
      _AdminVideoRequestApprovalRequestState();
}

class _AdminVideoRequestApprovalRequestState
    extends State<AdminVideoRequestApprovalRequest> {
  final FlutterSecureStorage storage = const FlutterSecureStorage();

  bool loading = true;
  List<String> statuses = [];
  Map<String, List<Map<String, dynamic>>> requestData = {};

  Future<Map<String, String>> _authHeaders() async {
    return {
      'Authorization': await storage.read(key: 'token') ?? '',
      'Admin-Token': await storage.read(key: 'adminToken') ?? '',
    };
  }

  Future<void> loadStatuses() async {
    final res = await http.get(
      Uri.parse('$server/admins/video_request_options/'),
      headers: await _authHeaders(),
    );

    final data = jsonDecode(res.body);
    statuses = List<String>.from(data['data'] ?? []);
  }

  Future<List<Map<String, dynamic>>> loadRequestsForStatus(
    String status,
  ) async {
    final res = await http.get(
      Uri.parse('$server/admins/video_requests/$status/'),
      headers: await _authHeaders(),
    );

    final data = jsonDecode(res.body)['data'] ?? {};
    return List<Map<String, dynamic>>.from(data.values);
  }

  Future<void> init() async {
    setState(() => loading = true);

    await loadStatuses();

    final Map<String, List<Map<String, dynamic>>> temp = {};
    for (final status in statuses) {
      temp[status] = await loadRequestsForStatus(status);
    }

    setState(() {
      requestData = temp;
      loading = false;
    });
  }

  Future<void> updateStatus(String serial, String newStatus) async {
    await http.patch(
      Uri.parse('$server/admins/review_video_request/$serial/'),
      headers: {...(await _authHeaders()), 'Content-Type': 'application/json'},
      body: jsonEncode({'request_status': newStatus}),
    );

    await init();
  }

  @override
  void initState() {
    super.initState();
    init();
  }

  @override
  Widget build(BuildContext context) {
    if (loading) {
      return const Center(child: CircularProgressIndicator());
    }

    return Scaffold(
      appBar: AppBar(title: const Text('Video Request Approval')),
      body: SingleChildScrollView(
        scrollDirection: Axis.horizontal,
        child: Row(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: statuses.map((status) {
            final requests = requestData[status] ?? [];

            return Container(
              width: 320,
              margin: const EdgeInsets.all(12),
              padding: const EdgeInsets.all(10),
              decoration: BoxDecoration(
                border: Border.all(color: Colors.grey.shade400),
                borderRadius: BorderRadius.circular(6),
              ),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(
                    status,
                    style: const TextStyle(
                      fontSize: 18,
                      fontWeight: FontWeight.bold,
                    ),
                  ),
                  const SizedBox(height: 10),

                  if (requests.isEmpty)
                    const Text('No requests in this status'),

                  ...requests.map((req) {
                    return Card(
                      margin: const EdgeInsets.symmetric(vertical: 8),
                      child: Padding(
                        padding: const EdgeInsets.all(8),
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            Text(
                              req['requeset_title'] ?? '',
                              style: const TextStyle(
                                fontWeight: FontWeight.bold,
                              ),
                            ),
                            Text('User: ${req['username']}'),
                            Text('Date: ${req['request_date']}'),
                            const SizedBox(height: 6),
                            Text(req['request_description'] ?? ''),
                            const SizedBox(height: 8),

                            DropdownButton<String>(
                              value: status,
                              isExpanded: true,
                              items: statuses
                                  .map(
                                    (s) => DropdownMenuItem(
                                      value: s,
                                      child: Text(s),
                                    ),
                                  )
                                  .toList(),
                              onChanged: (value) {
                                if (value != null && value != status) {
                                  updateStatus(req['serial'].toString(), value);
                                }
                              },
                            ),
                          ],
                        ),
                      ),
                    );
                  }),
                ],
              ),
            );
          }).toList(),
        ),
      ),
    );
  }
}
